### ISSUE TITLE
Prevent POI limits being reached when formatting columns

### ISSUE THREAD

**Issue by cfsimplicity** [cfsimplicity/lucee-spreadsheet#84]
See http://stackoverflow.com/questions/38343258/error-formating-a-lucee-spreadsheet

**Comment by cfprogrammer** [comment]
I am the person that posted this on Stack.  I was able to get by the original issue by using 'formatCellRange'.

I am now getting the same error 'The maximum number of cell styles was exceeded. You can define up to 4000 styles in a .xls workbook'  in a worksheet that has over 9000 rows whether or not I try to apply any styles at all.  I am using Lucee's number format function when I am sending the data to the spreadsheet, I don't know if that is invoking some type of built in style but it not even allowing me to create the spread sheet.

if there is anything I can share to help you debug please let me know.

**Comment by cfsimplicity** [comment]
Hi, I'm working on a fix right now which I hope to push shortly. Would be great if you could test if for me.

**Comment by cfsimplicity** [comment]
Could you post the code you're using with NumberFormat()?

**Comment by cfprogrammer** [comment]
I would be VERY happy to test. I have to leav in an hour but should be back a couple hours after.  happy to help

**Comment by cfprogrammer** [comment]
this is the code I am using to format the numbers

```

ss.addRow(workbook=arguments.wbObj, data=" | | | 
		                    |#numberFormat(grossCharge_fcTotal,'9.99')#
		                    |#numberFormat(contractualAllowance_fcTotal,'9.99')#
		                    |#numberFormat(netCharge_fcTotal,'9.99')#
		                    |#numberFormat(revenueAdjustment_fcTotal,'9.99')#
		                    |#numberFormat(paymentTotal_fcTotal,'9.99')#
		                    |#numberFormat(writeOff_fcTotal,'9.99')#
		                    |#numberFormat(refund_fcTotal,'9.99')#
		                    |#numberFormat(balanceDue_fcTotal,'9.99')#
		                    |#numberFormat(_0_30_fcTotal,'9.99')#
		                    |#numberFormat(_31_60_fcTotal,'9.99')#
		                    |#numberFormat(_61_90_fcTotal,'9.99')#
		                    |#numberFormat(_91_120_fcTotal,'9.99')#
		                    |#numberFormat(_121_150_fcTotal,'9.99')#
		                    |#numberFormat(_151_180_fcTotal,'9.99')#
		                    |#numberFormat(over_180_fcTotal,'9.99')#
		                    |#numberFormat(balanceDue_fcTotal,'9.99')#", delimiter="|");


```

**Comment by cfprogrammer** [comment]
if I didn't do this anything that was a 0 (zero) was being put into the spreadsheet as "FALSE"

**Comment by cfsimplicity** [comment]
So are you looping over that 9000+ times to create the workbook?

**Comment by cfprogrammer** [comment]
yes,
the customer wants a lot of empty rows and subtotals through out the worksheet

**Comment by cfprogrammer** [comment]
I can send you a sample of the final sheet if that is helpful.  Maybe there is a better way to do what I am trying to do

**Comment by cfsimplicity** [comment]
I think it's best to treat this as a separate issue with `addRow()`. I'll create a new ticket and look at it tomorrow or when I can. Meantime I'm going to push a fix soon for the `formatColumns()` function. If you could test that out I'd appreciate it. I'm in the UK so may not be up when you get back, but just leave a message here.

**Comment by cfprogrammer** [comment]
ok thanks for everything

**Comment by cfsimplicity** [comment]
I've pushed 0.7.5 which should now allow `formatColumns()` to work with your large workbook. If you confirm either way that would be great.

**Comment by cfprogrammer** [comment]
You maybe correct that it is the addRow instead.  If I limit the number of record passed to 3000 I am getting all 3 worksheets with formatting as I had hoped for.  But if I raise the limit of the records being returned I get the same error I have been seeing.

Thanks for working on this so quickly.

![image](https://cloud.githubusercontent.com/assets/12588948/16820973/ab9a0b20-4908-11e6-97bd-dff3cb7f7f8a.png)

**Comment by cfsimplicity** [comment]
Yes, cells of certain types are styled when you add data of that type using `addrow()`.

But could you just confirm whether `formatColumns()` is working for you so I can close this issue?

Then could you open up a new issue for `addRows` giving as much detail as you can. I've not been able to replicate the error you're seeing simply looping over `addRow()` and including values wrapped in `NumberFormat()`. I'm probably missing some aspect though.

Have to turn in now, but will have another look tomorrow.

**Comment by cfprogrammer** [comment]
It appears that I am no longer getting any errors from the formatColumn function.  Thank you for such a quick fix.

I will do some testing with the addRow function and let you know what I can figure out.

**Comment by cfsimplicity** [comment]
Thanks for confirming.

**Comment by cfsimplicity** [comment]
> if I didn't do this anything that was a 0 (zero) was being put into the spreadsheet as "FALSE"

I've fixed this too with version 0.7.7
